# SPARKS FOUNDATION INTERNSHIP TASK

## PREDICTION USING A REGRESSION MODEL

### AUTHOR :- SAMARTH JAIN

I have designed a simple regression model consisting of only 1 variable to predict the final score of a student based on the total hours they have studied using TuriCreate

In [1]:
import turicreate as tc
import pandas as pd

## importing the database using pandas and converting it to SFrame using turicreate

In [2]:
url = "http://bit.ly/w-data"
data = tc.SFrame(pd.read_csv(url))
print("Data imported successfully")

Data imported successfully


## materializing the database 

In [3]:
data.show()

Materializing SFrame

## complete database

In [4]:
data.print_rows(num_rows=25)

+-------+--------+
| Hours | Scores |
+-------+--------+
|  2.5  |   21   |
|  5.1  |   47   |
|  3.2  |   27   |
|  8.5  |   75   |
|  3.5  |   30   |
|  1.5  |   20   |
|  9.2  |   88   |
|  5.5  |   60   |
|  8.3  |   81   |
|  2.7  |   25   |
|  7.7  |   85   |
|  5.9  |   62   |
|  4.5  |   41   |
|  3.3  |   42   |
|  1.1  |   17   |
|  8.9  |   95   |
|  2.5  |   30   |
|  1.9  |   24   |
|  6.1  |   67   |
|  7.4  |   69   |
|  2.7  |   30   |
|  4.8  |   54   |
|  3.8  |   35   |
|  6.9  |   76   |
|  7.8  |   86   |
+-------+--------+
[25 rows x 2 columns]



## creating a training set and a test set using random split

In [5]:
training_set, test_set = data.random_split(.8,seed=0)


## training a basic regression model using the training set with Hours as features and Scores as the target

In [6]:
reg_model = tc.linear_regression.create(training_set,target='Scores',features=['Hours'])

Linear regression:

--------------------------------------------------------

Number of examples          : 23

Number of features          : 1

Number of unpacked features : 1

Number of coefficients    : 2

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------------------------+

| Iteration | Passes   | Elapsed Time | Training Max Error | Training Root-Mean-Square Error |

+-----------+----------+--------------+--------------------+---------------------------------+

| 1         | 2        | 1.003168     | 9.858140           | 5.346648                        |

+-----------+----------+--------------+--------------------+---------------------------------+

SUCCESS: Optimal solution found.

## testing the model on the test set

In [7]:
x = test_set
y = reg_model.predict(test_set)
df = pd.DataFrame({'Actual': x, 'Predicted': y})  
df 

Actual  Predicted
0  {'Hours': 4.5, 'Scores': 41}  46.545354
1  {'Hours': 8.9, 'Scores': 95}  88.689419

## displaying the coefficients of the features used in the regression model

In [8]:
reg_model.coefficients

Columns:
	name	str
	index	str
	value	float
	stderr	float

Rows: 2

Data:
+-------------+-------+--------------------+--------------------+
|     name    | index |       value        |       stderr       |
+-------------+-------+--------------------+--------------------+
| (intercept) |  None | 3.4434702173310683 | 2.600912865430965  |
|    Hours    |  None | 9.578196470074936  | 0.4777871747388556 |
+-------------+-------+--------------------+--------------------+
[2 rows x 4 columns]

## using the model to predict the score of a student who has studied 9.25 hours

In [9]:
hours = ({'Hours' : 9.25})

ans = reg_model.predict(hours)

print ('A student studying for 9.25 hours will score',ans,'% based on this regression model')

A student studying for 9.25 hours will score [92.04178756552422] % based on this regression model


### the model predicts that a student who has studied 9.25 hours will score 92%